### PR TITLE
[Frontend] New `allowed_token_ids` decoding request parameter

### PR DESCRIPTION
**Pull request by njhill**
This PR adds a new `LogitsProcessor` for constraining decoded tokens to a fixed set of token ids. This is needed for some of our classification use cases.

It is exposed via a new `allowed_token_ids` parameter in the OpenAI completion API (wouldn't be applicable to chat use cases).

### CODE REVIEW

**Comment by DarkLight1337** [comment]
This reminds me of #5986...

File: vllm/entrypoints/openai/logits_processors.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from typing import Dict, FrozenSet, Iterable, List, Optional, Union
+
+import torch
+from transformers import PreTrainedTokenizer
+
+from vllm.sampling_params import LogitsProcessor
+
+
+class AllowedTokenIdsLogitsProcessor:
+    """Logits processor for constraining generatedtokens to a
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/logits_processors.py:11]:
```suggestion
    """Logits processor for constraining generated tokens to a
```

File: vllm/entrypoints/openai/protocol.py [hunk 2]
```diff
@@ -5,9 +5,12 @@
 
 import torch
 from pydantic import BaseModel, ConfigDict, Field, model_validator
+from transformers import PreTrainedTokenizer
+# pydantic needs the TypedDict from typing_extensions
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/protocol.py:9]:
```suggestion

```

**Review by DarkLight1337 — APPROVED** [review]
Overall LGTM, just some minor nits.

File: vllm/entrypoints/openai/logits_processors.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    specific set of token ids."""
+
+    def __init__(self, allowed_ids: Iterable[int]):
+        self.allowed_ids: Optional[List[int]] = list(allowed_ids)
+        self.mask: Optional[torch.Tensor] = None
+
+    def __call__(self, token_ids: List[int],
+                 logits: torch.Tensor) -> torch.Tensor:
+        if self.mask is None:
+            self.mask = torch.full((logits.shape[-1], ),
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/logits_processors.py:21]: Maybe can use `torch.ones` with boolean type?

**Comment by njhill** [comment]
@DarkLight1337 it's a bit different in that this would be used in conjunction with a limited set of specific class tokens and would probably only be used with max_tokens=1, it's not intended for text generation. #5986 is for excluding certain words from generated text output.

File: vllm/entrypoints/openai/logits_processors.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                                   True,
+                                   dtype=torch.bool,
+                                   device=logits.device)
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/logits_processors.py:24]: Is it guaranteed that logits will always be from the same device even for multi-gpu inference?

**Comment by njhill** [comment]
Thanks @DarkLight1337, I'll add a test too.

File: vllm/entrypoints/openai/logits_processors.py (hunk 4 above)
Comment (njhill) [reply vllm/entrypoints/openai/logits_processors.py:23]: @DarkLight1337 yes I think this is safe to assume. There's never more than one GPU device used per process AFAIK.

**Comment by mgoin** [comment]
Could the same result be achieved with high logit_bias on the token ids?

**Comment by njhill** [comment]
@mgoin we thought about this but it's not really reliable since we don't know what the logits distribution looks like .. what value do we use in a general case to ensure that the tokens of interest all end up higher than any other token, while not overflowing

**Comment by whyiug** [comment]
Why isn’t chat case applicable? 
Additionally, how can this be used with multimodal models? I've encountered some classification cases involving multimodal models. Appreciate your guidance. @njhill
